guides/internals.rst

```diff
@@ -0,0 +1,321 @@
+
+To run a specific test file manually, you can use pytest. I usually use the
+following invocation:
+
+.. code-block::
+
+    python -m pytest tests/cover/test_conjecture_engine.py
+
+You will need to have Hypothesis installed locally to run these. I recommend a
+virtualenv where you have run ``pip install -e .``, which installs all the
+dependencies and puts your ``src`` directory in the path of installed packages
+so that edits you make are automatically pipped up.
+
+Useful arguments you can add to pytest are `` -n 0``, which will disable build
+parallelism (I find that on my local laptop the startup time is too high to be
+worth it when running single files, so I usually do this), and `` -kfoo`` where
+foo is some substring common to the set of tests you want to run (you can also
+use composite expressions here. e.g. `` -k'foo and not bar'`` will run anything
+containing foo that doesn't also contain bar).
+
+-----------------------
+Engine Design Specifics
+-----------------------
+
+There are a couple of code patterns that are mostly peculiar to Conjecture that
+you may not have encountered before and are worth being aware of.
+
+~~~~~~~~~~~~~~~~~~~~
+Search State Objects
+~~~~~~~~~~~~~~~~~~~~
+
+There are a number of cases where we find ourself with a user-provided function
+(where the "user" might still be something that is entirely our code) and we
+want to pass a whole bunch of different examples to it in order to achieve some
+result. Currently this includes each of the main engine, the Shrinker (in
+engine.py) and the minimizer, but there are likely to be more in future.
+
+We typically organise such things in terms of an object that you create with
+the function and possibly an initial argument that stores these on self and
+has some ``run`` or similar method. They then run for a while, repeatedly
+calling the function they were given.
+
+Generally speaking they do not call the function directly, but instead wrap
+calls to it. This allows them to implement a certain amount of decision caching,
+e.g. avoiding trying the same shrink twice, but also gives us a place where we
+can update metadata about the search process.
+
+~~~~~~~~~~~
+Weird Loops
+~~~~~~~~~~~
+
+The loops inside a lot of the engine look very strange and unidiomatic. For
+example:
+
+.. code-block:: python
+
+        i = 0
+        while i < len(self.intervals):
+            u, v = self.intervals[i]
+            if not self.incorporate_new_buffer(
+                self.shrink_target.buffer[:u] + self.shrink_target.buffer[v:]
+            ):
+                i += 1
+
+
+The more natural way to write this in Python would of course be:
+
+.. code-block:: python
+
+        for u, v in self.intervals:
+            self.incorporate_new_buffer(
+                self.shrink_target.buffer[:u] + self.shrink_target.buffer[v:]
+            )
+
+This way of writing the loop would be *entirely wrong*.
+
+Every time `incorporate_new_buffer` succeeds, it changes the shape of the
```

Comment:
nit: needs double backticks

Reply:
nit: actually failures to use RST correctly are things that need fixing rather than merely nits. ;-)